### PR TITLE
Added Deserialize and compatible Serialize

### PR DESCRIPTION
**Pull request by fwiesel**
To transport SOAP objects over other transport mechanisms,
such as oslo.messaging, we need a representation,
which allows us to reconstruct (de-serialize) the
serialized object. This is commonly a dict, as done in
the pre-existing serialize_object function.

This change improves the function by:
1. Not serializing empty defaults making the result
   less verbose
2. Serialize the type by adding the "class" field to
   the dict, in cases of fields being of derived types
   of the default

For compatibility reasons, the latter is only active
when opting-in.

Additionally, the deserialize_object has been added,
which takes the dict generated by serialize_object
and creates a SOAP object out of it

Change-Id: Ie76b1e6940b5022563ce91d5692df589573704d0

### CODE REVIEW

**Comment by fwiesel** [comment]
Essentially what we already have in Nova.
https://github.com/sapcc/nova/blob/stable/rocky-m3/nova/virt/vmwareapi/vim_util.py#L153-L238

File: oslo_vmware/vim_util.py [hunk 1]
```diff
@@ -679,24 +679,118 @@ def is_vim_instance(o, vim_type_name):
...
-    """Convert Suds object into a dict."""
+def _get_child_type(sxtype, name):
+    if not sxtype:
+        return None
+    child, _ = sxtype.get_child(name)
+    return child.type[0]
+
+
+def serialize_object(obj, typed=False, expected_type=None):
+    """Serialize an SOAP object to a dict
```
Comment (joker-at-work) [inline oslo_vmware/vim_util.py:690]: nitpick: Should be `a` not `an`, right?

File: oslo_vmware/vim_util.py [hunk 2]
```diff
@@ -679,24 +679,118 @@ def is_vim_instance(o, vim_type_name):
...
+                d[k] = ser
         else:
-            d[k] = v
+            if not typed or v is not None:
+                d[k] = v
     return d
 
 
+def deserialize_object(factory, serialized, typename):
+    """"Deserialize an dict to a SOAP object
```
Comment (joker-at-work) [inline oslo_vmware/vim_util.py:751]: nitpick: `a dict`

**Review by joker-at-work — APPROVED** [review]
Would you mind adding a little test for that, especially for the special cases, so that we can see that deserialize works with the output of serialize?

File: oslo_vmware/vim_util.py [hunk 3]
```diff
@@ -679,24 +679,118 @@ def is_vim_instance(o, vim_type_name):
...
+    child, _ = sxtype.get_child(name)
+    return child.type[0]
+
+
+def serialize_object(obj, typed=False, expected_type=None):
+    """Serialize an SOAP object to a dict
+
+    For the function to serializes it in a format that the corresponding
+    deserialize_object can restore the original object, it needs to get
+    get passed typed=True, and the expected_type as string for polymorphism.
```
Comment (leust) [inline oslo_vmware/vim_util.py:694]: A few typos here, too. `to serialize`, `get get`.

File: oslo_vmware/vim_util.py (hunk 2 above)
Comment (fwiesel) [reply oslo_vmware/vim_util.py:751]: Fixed

File: oslo_vmware/vim_util.py (hunk 1 above)
Comment (fwiesel) [reply oslo_vmware/vim_util.py:690]: Correct, fixed it.

File: oslo_vmware/vim_util.py (hunk 3 above)
Comment (fwiesel) [reply oslo_vmware/vim_util.py:694]: Fixed

**Review by leust — APPROVED** [review]

**Pull request merged**